pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,240 @@
+        throw new AssertionError("Can't instantiate utility class");
+    }
+    
+    /**
+     * Starts global tracking. Allows tracking operations to take place and starts the wall clock.
+     * Must be called once PMD starts if tracking is desired, no tracking will be performed otherwise.
+     */
+    public static void startGlobalTracking() {
+        wallClockStart = System.currentTimeMillis();
+        trackTime = true;
+        initThread(); // init main thread
+    }
+    
+    /**
+     * Stops global tracking. Stops the wall clock. All further operations will be treated as NOOP.
+     * @return The timed data obtained through the run.
+     */
+    public static TimingReport stopGlobalTracking() {
+        if (!trackTime) {
+            return null;
+        }
+        
+        finishThread(); // finish the main thread
+        trackTime = false;
+        
+        // Fix UNACCOUNTED metric (total time is meaningless)
+        final TimedResult unaccountedResult = ACCUMULATED_RESULTS.get(
+                new TimedOperation(TimedOperationCategory.UNACCOUNTED, null));
+        unaccountedResult.totalTime.set(unaccountedResult.selfTime.get());
+        
+        return new TimingReport(System.currentTimeMillis() - wallClockStart, ACCUMULATED_RESULTS);
+    }
+    
+    /**
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static void initThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get(); // Just make sure it's initialized
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().size() == 1) {
+            finishOperation();
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation
+     * @param category The category under which to track the operation.
+     */
+    public static void startOperation(final TimedOperationCategory category) {
+        startOperation(category, null);
+    }
+    
+    /**
+     * Starts tracking an operation
+     * @param category The category under which to track the operation.
+     * @param label A label to be added to the category. Allows to differentiate measures within a single category.
+     */
+    public static void startOperation(final TimedOperationCategory category, final String label) {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get().add(new TimerEntry(category, label));
+    }
+    
+    /**
+     * Finishes tracking an operation
+     */
+    public static void finishOperation() {
+        finishOperation(0);
+    }
+    
+    /**
+     * Finishes tracking an operation
+     * @param extraDataCounter An optional additional data counter to track along the measurements
```

Comment:
Maybe it would be nice to add an example of what this counter can be used for. From what I've seen, it's used to track the number of evaluations of a rulechain rule -> could be more explicit